IDN Table Support

wbond opened the pull request:
Implementation based on discussion at https://github.com/zonedb/zonedb/issues/174

ydnar commented:
This is amazing.

Does it update the zone codetable for all languages from the individual IDN tables?

wbond commented:
No, but that should probably happen. I'll get that added tomorrow morning.

wbond commented:
I've updated the PR with functionality to update `Zone.CodePoints` from the IDN tables. In the process I fixed a number of bugs with `CodeTable.Compress()` and added tests for it.

cee-dub commented on build/unicode_test.go:
```diff
@@ -0,0 +1,98 @@
+
+	table.AppendRune(95)
+	CompareTable(t, *table, "09--,,__", false)
+}
```
These tests are great. I think they'd be even easier to understand if the used the notation for the rune values of characters `'0'`, rather than the values directly `48` (which makes me go look at an ASCII table).

wbond replied:
I agree, I'll get that tweaked either later tonight, or tomorrow morning.

cee-dub replied:
Maybe that's just me, I don't want to block this from getting in.

ydnar replied:
Agree, using rune literals will make this easier to read. Happy Go supports UTF-8 in source.

ydnar commented on build/unicode_test.go:
```diff
@@ -0,0 +1,98 @@
+	CompareTable(t, *table, "0912/18; @", false)
+}
+
+func TestAppendRangeCompress(t *testing.T) {
```
This is awesome.

wbond commented:
I updated the tests to use rune literals.

Additionally, I added a new commit with two new methods on `Zone`: `Zone.IsValidDomain()` and `Zone.IDNTable()`.

ydnar commented on zone.go:
```diff
@@ -115,6 +120,79 @@ func (z *Zone) IsInRootZone() bool {
+
+var asciiCodePoints = []rune{'-', '-', '0', '9', 'a', 'z'}
+
+func domainInPoints(domain string, codePoints []rune) bool {
```
Rename to `stringInCodePoints`.

ydnar commented on zone.go:
```diff
@@ -115,6 +120,79 @@ func (z *Zone) IsInRootZone() bool {
+	return true
+}
+
+func charInPoints(char rune, codePoints []rune) bool {
```
First, rename to `runeInCodePoints`. Rename args `char` -> `c` and `codePoints` -> `points`.

Second, let’s optimize this by recursively splitting `points` and testing against outer bounds.

1. Assuming the code points are sorted, then we can test `c >= points[0]` and `c <= points[len(points)-1]`.
2. Then split `points` around a pivot in the middle, and recursively call `runeInCodePoints` (or iterate, rather than recurse) until we prove or disprove `c` is within the range of code points.

Go has a couple examples of similar code here: https://github.com/golang/go/blob/master/src/unicode/letter.go#L125

cee-dub commented on zone.go:
```diff
@@ -115,6 +120,79 @@ func (z *Zone) IsInRootZone() bool {
 	return z.IsTLD() && z.IsDelegated()
 }
 
+// Checks to see if a domain is valid according to character set restrictsion
```
sp. restrictsion

wbond commented:
I revised the last commit based on your feedback, thanks!

I ended up writing some benchmarks since I noticed the linked `letter.go` switched between linear and binary search algorithms based on the number of ranges to search through.

Unfortunately it seems the extra comparisons in my implementation of binary search wipe out the benefit we should get from the algorithm improvement. I tried to vary the test data some, with both matching and non-matching ASCII and unicode searched against a set of points for ASCII, cyrillic languages and then a big list of unicode capital letters.

```
PASS
BenchmarkManyUnicodeLinearRuneInCodePoints-8	 3000000	       545 ns/op
BenchmarkManyUnicodeBinaryRuneInCodePoints-8	 1000000	      1135 ns/op
BenchmarkUnicodeLinearRuneInCodePoints-8    	10000000	       124 ns/op
BenchmarkUnicodeBinaryRuneInCodePoints-8    	10000000	       167 ns/op
BenchmarkAsciiLinearRuneInCodePoints-8      	10000000	       116 ns/op
BenchmarkAsciiBinaryRuneInCodePoints-8      	10000000	       131 ns/op
```

Let me know if I botched the binary search, or of any improvements I should to the implementation of it.

cee-dub commented:
At 545 ns/op, comparing the worst case domain label length 63 is only going to add ~3ms. I think this is fast enough.

ydnar commented on zone.go:
```diff
@@ -109,6 +120,133 @@ func (z *Zone) IsInRootZone() bool {
+
+func linearRuneInCodePoints(c rune, points []rune) bool {
+	i := 0
+	num := len(points)
```
`num` is used twice as `num-1`, perhaps replace `num` with `end`?

ydnar commented on zone.go:
```diff
@@ -109,6 +120,133 @@ func (z *Zone) IsInRootZone() bool {
+	i := 0
+	num := len(points)
+
+	if c < points[0] || c > points[num-1] {
```
Check `c > points[num-1]` first, because it’s more likely.

ydnar commented on zone.go:
```diff
@@ -109,6 +120,133 @@ func (z *Zone) IsInRootZone() bool {
+}
+
+func stringInCodePoints(domain string, points []rune) bool {
+	for _, c := range domain {
```
At the expense of scanning the string twice, can we find the min/max `rune` in the `domain` string and test if both are `runeInCodePoints`?

ydnar commented on zone.go:
```diff
@@ -109,6 +120,133 @@ func (z *Zone) IsInRootZone() bool {
+
+var asciiCodePoints = []rune{'-', '-', '0', '9', 'a', 'z'}
+
+func stringsInCodePoints(chunks []string, points []rune) bool {
```
`labelsInCodePoints`?

wbond commented:
@ydnar A bunch of your comments come through not attached to line numbers, FYI.

wbond commented:
Could you clarify where you were thinking of placing a `<=`?

wbond commented:
Using bit shifts made the biggest improvement in binary vs linear search, but linear is still faster.

```
BenchmarkManyUnicodeLinearRuneInCodePoints-8	 3000000	       526 ns/op
BenchmarkManyUnicodeBinaryRuneInCodePoints-8	 2000000	       885 ns/op
BenchmarkUnicodeLinearRuneInCodePoints-8    	10000000	       109 ns/op
BenchmarkUnicodeBinaryRuneInCodePoints-8    	10000000	       141 ns/op
BenchmarkAsciiLinearRuneInCodePoints-8      	10000000	       106 ns/op
BenchmarkAsciiBinaryRuneInCodePoints-8      	10000000	       116 ns/op
```

My guess is our current storage strategy of using a slice of runes alternating lo, hi is affecting performance since we need to effectively floor the calculation by a divide multiply. If it we a slice of a struct with two fields, it may improve performance some.

Looping through the string to find the min and max runes did not improve performance at all, but generally had a very slight hit in performance for everything but a linear search of a short list of code points:

```
BenchmarkManyUnicodeLinearRuneInCodePoints-8	 3000000	       536 ns/op
BenchmarkManyUnicodeBinaryRuneInCodePoints-8	 2000000	       942 ns/op
BenchmarkUnicodeLinearRuneInCodePoints-8    	20000000	       107 ns/op
BenchmarkUnicodeBinaryRuneInCodePoints-8    	10000000	       147 ns/op
BenchmarkAsciiLinearRuneInCodePoints-8      	20000000	       105 ns/op
BenchmarkAsciiBinaryRuneInCodePoints-8      	10000000	       117 ns/op
```

wbond commented:
I've included the tweaks you suggested @ydnar, and rebased on master.

I'm thinking we should take out the min/max rune detection in `stringInCodePoints` since it affects the pathological (slowest) case the most. For common checks, the difference seems to be ~1ns using the linear algorithm.

ydnar commented on zone.go:
```diff
@@ -109,6 +120,154 @@ func (z *Zone) IsInRootZone() bool {
+		}
+	}
+
+	return "", errors.New("domain is not a valid IDN of the zone")
```
This should be a predefined ErrIDNTableNotFound or similar.

ydnar commented:
Ok. This is as good as we're going to get. Thanks @wbond!

I had trouble parsing recent changes. Did you rebase or force push?

Second, I noticed something in the overall logic we might want to change in a future PR: Validity and table detection is subjective. For instance, while ICANN and the registry might prevent certain Unicode chars at the immediate subdomain of a Zone, I'm not sure this applies to hostnames.

E.g. 🍻.domainr.com might very well be possible, despite com not permitting emoji at the 2nd level.

My instinct is to rip out all the labels, IDN, splitting, and validity checks out of zonedb and replace with a simple API:

Zone.IDNTable(unicodeLabel string) (string, error)

CodePoints.ContainsRune(c rune) bool

Or something.

ydnar commented:
Last comment implies:

type CodePoints []rune

And maybe:

type IDNTable struct {
    Lang string
    CodePoints
}

wbond commented on zone.go:
```diff
@@ -115,6 +120,154 @@ func (z *Zone) IsInRootZone() bool {
+	}
+	if !runeInCodePoints(min, points) {
+		return false
+	}
```
Since lines 202-219 slow down the worse case, but don't really improve the best case, they should be removed, correct?

ydnar replied:
Sure.

wbond commented on zone.go:
```diff
@@ -115,6 +120,154 @@ func (z *Zone) IsInRootZone() bool {
+	return linearRuneInCodePoints(c, points)
+}
+
+func binaryRuneInCodePoints(c rune, points []rune) bool {
```
I presume I should remove this implementation and related benchmarks since it is slower than the linear search?

ydnar replied:
@wbond let’s leave it and the benchmarks in. I still think there’s an opportunity to refine this.
